types/wlr_surface.c

```diff
@@ -170,9 +170,17 @@ static void surface_state_viewport_src_size(struct wlr_surface_state *state,
 static void surface_state_finalize(struct wlr_surface *surface,
 		struct wlr_surface_state *state) {
 	if ((state->committed & WLR_SURFACE_STATE_BUFFER)) {
-		if (state->buffer_resource != NULL) {
-			wlr_resource_get_buffer_size(state->buffer_resource,
-				&state->buffer_width, &state->buffer_height);
+		if (state->buffer_resource) {
+			state->buffer = wlr_buffer_from_resource(surface->renderer,
```

Comment:
When we overwrite a `state->buffer` field, we need to make sure it's NULL. If it's not, it means it's holding a `wlr_buffer` reference, so overwriting it will leak a reference.